middlewares/passport-client-cert.js

```diff
@@ -0,0 +1,111 @@
+ * client certificate credentials submitted in the TLS handshake
+ *
+ * Applications must supply a `verify` callback which accepts a `username` and
+ * an `info` object, which includes the certificate `fingerprint`. It then
+ * calls the `done` callback supplying a `user`. User should be set to `false`
+ * if the credentials are not valid.  If an exception occured, `err` should be
+ * set.
+ *
+ * Options:
+ *   - `passReqToCallback`  when `true`, `req` is the first argument to the
+ *      verify callback (default: `false`)
+ *
+ * Examples:
+ *
+ * passport.use(new ClientCertStrategy(
+ *   function (userName, info, done) {
+ *     if (!config.auth.client_certificates_enabled) {
+ *       return done(new UnauthorizedError('Unsupported authentication method'))
+ *     }
+ *
+ *     Account.findByName(userName)
+ *       .then(function (userObj) {
+ *         if (!userObj || userObj.is_disabled || userObj.fingerprint !== info.fingerprint) {
+ *           return done(new UnauthorizedError('Unknown or invalid account'))
+ *         }
+ *         done(null, userObj)
+ *       })
+ *   }))
+ *
+ *
+ * @param {Object} options
+ * @param {Function} verify
+ * @api public
+ */
+class ClientCertStrategy extends Strategy {
+  constructor (options, verify) {
+    if (typeof options === 'function') {
+      verify = options
+      options = {}
+    }
+    if (!verify) throw new Error('Client cert authentication strategy requires a verify function')
+
+    super()
+
+    this.name = 'client-cert'
+    this._verify = verify
+    this._passReqToCallback = options.passReqToCallback
+  }
+
+  _verified (err, user) {
+    if (err) { return this.error(err) }
+    if (!user) { return this.fail() }
+  }
+
+  authenticate (req, options) {
+    // Requests must be authorized
+    // (i.e. the certificate must be signed by at least one trusted CA)
+    if (!req.socket.authorized) {
+      this.fail()
+      return
+    }
+
+    // { subject:
+    //    { C: 'US',
+    //      ST: 'MA',
+    //      L: 'Boston',
+    //      O: 'Example Co',
+    //      OU: 'techops',
+    //      CN: 'client1',
+    //      emailAddress: 'certs@example.com' },
+    //   issuer:
+    //    { C: 'US',
+    //      ST: 'MA',
+    //      L: 'Boston',
+    //      O: 'Example Co',
+    //      OU: 'techops',
+    //      CN: 'ca',
+    //      emailAddress: 'certs@example.com' },
+    //   fingerprint: 'E5:F6:7F:34:6B:28:92:9C:07:18:0B:46:5A:D5:E4:50:CA:4F:DD:25',
+    const clientCert = req.socket.getPeerCertificate()
+
+    if (!clientCert || !clientCert.subject.CN) {
+      this.fail()
+      return
+    }
+
+    // The CN is used as the username
+    const userName = clientCert.subject.CN
+
+    const info = {
+      // fingerprint used for certificate pinning
+      fingerprint: clientCert.fingerprint
+    }
+
+    if (this._passReqToCallback) {
+      this._verify(req, userName, info, this._verified.bind(this))
```

Comment:
@justmoon based on the discussion earlier of common name vs fingerprint, maybe it would be best to pass back an object with `commonName` and `fingerprint` here, as this would make the middleware implementation most flexible